shielder/contract/lib.rs

```diff
@@ -0,0 +1,93 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+mod merkle;
+mod psp22;
+mod types;
+mod errors;
+mod mocked_zk;
+
+#[ink::contract]
+mod contract {
+
+    use ink::storage::Mapping;
+
+    use crate::{merkle::MerkleTree, psp22::PSP22, types::{Set, Scalar}, errors::ShielderError, mocked_zk::{ZkProof, self}};
+
+    #[derive(Clone, Copy, scale::Encode, scale::Decode)]
+    #[cfg_attr(
+        feature = "std", 
+        derive(scale_info::TypeInfo)
+    )]
+    pub enum OpPub {
+        Deposit(u128, AccountId, AccountId),
+        Withdraw(u128, AccountId, AccountId)
+    }
+
+    #[ink(storage)]
+    pub struct Contract {
+        nullifier_set: Set<Scalar>,
+        notes: MerkleTree,
+    }
+
+    impl Contract {
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Self {
+                nullifier_set: Mapping::new(),
+                notes: MerkleTree::new()
+            }
+        }
+
+        #[ink(message)]
+        pub fn add_note(&mut self, op_pub: OpPub, h_note_new: Scalar) -> Result<(), ShielderError>{
```

Comment:
If this is just about adding a new, fresh note, then it shouldn't take `op_pub` as input. It's just initializing the account then.